File: cs/Scenarios/Server.NodeAccess/NodeManager.cs
// Copyright (c) Traeger Industry Components GmbH. All Rights Reserved.

namespace NodeAccess
{
    using System.Collections.Generic;

    using Opc.UaFx;
    using Opc.UaFx.Server;

    internal partial class NodeManager : OpcNodeManager
    {
        #region ---------- Public constructors ----------

        public NodeManager()
            : base("http://sampleserver/nodeaccess")
        {
        }

        #endregion

        #region ---------- Public properties ----------

        public OpcAccessControlList AccessControl
        {
            get;
            set;
        }

        #endregion

        #region ---------- Internal methods ----------

        internal bool CanWrite(OpcVariableNode node, OpcUserIdentity sessionIdentity)
        {
            var entries = this.AccessControl.Entries;

            foreach (var entry in entries) {
                if (entry.Principal.Identity is SystemIdentity serverIdentity) {
                    if (serverIdentity.DisplayName == sessionIdentity.DisplayName) {
                        return entry.IsAllowed(OpcRequestType.Write)
                                && serverIdentity.IsAllowed(node.Id.ValueAsString);
                    }
                }
            }

            return false;
        }

        #endregion

        #region ---------- Protected methods ----------

        protected override IEnumerable<IOpcNode> CreateNodes(OpcNodeReferenceCollection references)
        {
            var machineNode = new OpcObjectNode(this.DefaultNamespace.GetName("Machine"));

            new VariableNode<int>(this, machineNode, "Status", value: 42);
            new VariableNode<string>(this, machineNode, "Job", value: "JOB-0543");

            new VariableNode<bool>(this, machineNode, "Shutdown", value: false);
            new VariableNode<short>(this, machineNode, "Speed", value: 1000);
            new VariableNode<string>(this, machineNode, "Tooling", value: "cutter");

            references.Add(machineNode, OpcObjectTypes.ObjectsFolder);
            return new IOpcNode[] { machineNode };
        }

        #endregion
    }
}

File: vb/Scenarios/Server.NodeAccess/NodeManager.vb
' Copyright (c) Traeger Industry Components GmbH. All Rights Reserved.

Imports Opc.UaFx
Imports Opc.UaFx.Server

Namespace NodeAccess
    Partial Friend Class NodeManager
        Inherits OpcNodeManager

        ' ---------- Public constructors ----------

        Public Sub New()
            MyBase.New("http://sampleserver/nodeaccess")
        End Sub

        ' ---------- Public properties ----------

        Public Property AccessControl As OpcAccessControlList

        ' ---------- Internal methods ----------

        Friend Function CanWrite(ByVal sessionIdentity As OpcUserIdentity) As Boolean
            Dim entries = Me.AccessControl.Entries
            Dim serverIdentity As OpcUserIdentity = Nothing

            For Each entry In entries
                serverIdentity = TryCast(entry.Principal.Identity, OpcUserIdentity)

                If (serverIdentity IsNot Nothing) Then
                    If (serverIdentity.DisplayName = sessionIdentity.DisplayName) Then
                        Return entry.IsAllowed(OpcRequestType.Write)
                    End If
                End If
            Next

            Return False
        End Function

        ' ---------- Protected methods ----------

        Protected Overrides Function CreateNodes(ByVal references As OpcNodeReferenceCollection) As IEnumerable(Of IOpcNode)
            Dim folderNode = New OpcFolderNode(Me.DefaultNamespace.GetName("Folder"))

            Dim var01 = New VariableNode(Of Integer)(Me, folderNode, "Var01", value:=42)
            Dim var02 = New VariableNode(Of String)(Me, folderNode, "Var02", value:="Hello World")

            references.Add(folderNode, OpcObjectTypes.ObjectsFolder)
            Return New IOpcNode() {folderNode}
        End Function
    End Class
End Namespace
